pkg/asset/installconfig/aws/subnet.go

```diff
@@ -0,0 +1,165 @@
+package aws
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+var vpc string
+var vpcFromSubnet string
+var routeTables []*ec2.RouteTable
+var subnetMetadata map[string]*Subnet
+
+// Subnet hold metadata for a subnet.
+type Subnet struct {
+	// ARN is the subnet's Amazon Resource Name.
+	ARN string
+
+	// Public is true if the cloud-provider logic considers the subnet
+	// to be public (e.g. with suitable routing for hosting public
+	// load balancers).
+	Public bool
+
+	// VPC is the ID of the VPC to which the subnet belongs.
+	VPC string
+
+	// Zone is the subnet's availability zone.
+	Zone string
+}
+
+// SubnetMetadata retrieves metadata for the given subnet(s).  Results
+// are cached, so additional requests for a given subnet do not result
```

Comment:
there's no lock to prevent updates to global variables.

Reply:
Lack of locking and other shortcuts discussed in (and hopefully motivated by ;) 04215b96009d7c763961f2daaa9b6883520096ff .

Reply:
> Because it gets used in a few places, and because AWS API calls are
expensive, the metadata-fetcher has package-scoped globals for caching
previous results.

how are we measuring this? These are all GET/LIST calls which i don't think is that heavy. Also using subnets cuts our API calls later on drastically.. 

> We could cache them in a new asset, but we'll need
these results to verify the install-config asset, and we don't want to
store this AWS-API-provided information in the install config itself.

why not create a new asset that can store this once as this information is closely tied to install-config.yaml ?

> But it should be sufficient for the
installer-bring-your-own-subnets use case, and we can revisit later if
we need something more polished.

I would make the case that least polished is asking users for private and public subnets ;) and as we are taking on complexity for letting users git single list, we take on some complexity for filter them..

Reply:
> These are all GET/LIST calls which i don't think is that heavy.

Asking for them four times during a single asset build is so ugly :p.  But I can reroll to do that if you won't accept a caching cludge.

> why not create a new asset that can store this once...

Because that asset would depend on install-config for the user-input values, and install-config would depend on the new asset for validation metadata.  But maybe it could go on the install-config asset and somehow not get persisted to disk...

> I would make the case that least polished is asking users for private and public subnets...

Internal polishing is easy.  API-touching polishing is a pain ;).

> ... and as we are taking on complexity for letting users git single list, we take on some complexity for filter them.

We'd need that complexity to validate user-supplied separate list too.  I don't think it's avoidable.

Reply:
I've filed #2512 with preliminary metadata handling on `InstallConfig`.  I'll rebase this PR on top once that lands and gives a more elegant place to store cached values than this PR's current package globals.